src/borrow_tracker/tree_borrows/tree/tests.rs

```diff
@@ -75,151 +75,109 @@ fn protected_enforces_noalias() {
-///
-/// `AccessRel2YChildX` could also theoretically exist, but we are
-/// not interested in inserting spurious reads through `x` in that case.
-trait Into2Rel: Clone + Copy + Exhaustive + fmt::Display {
-    /// Message to display to explain the relative position of `x` and `y`.
-    const LEGEND: &'static str;
-    /// Variant to emit for an access directly through `x` or `y`.
-    const THROUGH_X: Self;
-    const THROUGH_Y: Self;
-    /// Determines if it is possible for `y` not to exist while `x` is being
-    /// retagged, given their relative position.
-    const ALLOWS_X_RETAG_FIRST: bool;
-    /// A relative position to a pair `w rel (x, y)` can be converted to a pair
-    /// of relative positions `(w rel x, w rel y)`.
-    /// In general not all combinations of output pairs are possible.
-    fn into_pair(self) -> (AccessRelatedness, AccessRelatedness);
-}
-
-impl Into2Rel for AccessRel2MutuallyForeign {
-    const LEGEND: &'static str = "x and y are mutually foreign";
-    const THROUGH_X: Self = Self::ChildX;
-    const THROUGH_Y: Self = Self::ChildY;
-    const ALLOWS_X_RETAG_FIRST: bool = true;
-    fn into_pair(self) -> (AccessRelatedness, AccessRelatedness) {
+impl RelPosXY {
+    fn allows_x_retag_first(self) -> bool {
         match self {
-            AccessRel2MutuallyForeign::ChildX =>
-                (AccessRelatedness::StrictChildAccess, AccessRelatedness::DistantAccess),
-            AccessRel2MutuallyForeign::ChildY =>
-                (AccessRelatedness::DistantAccess, AccessRelatedness::StrictChildAccess),
-            AccessRel2MutuallyForeign::Other =>
-                (AccessRelatedness::DistantAccess, AccessRelatedness::DistantAccess),
+            RelPosXY::MutuallyForeign => true,
+            RelPosXY::XChildY => false,
         }
     }
 }
 
-impl Into2Rel for AccessRel2XChildY {
-    const LEGEND: &'static str = "x is a strict child of y";
-    const THROUGH_X: Self = Self::ChildBoth;
-    const THROUGH_Y: Self = Self::ChildY;
-    // Parent `y` has to already exist upon creation of `x` its child.
-    const ALLOWS_X_RETAG_FIRST: bool = false;
-    fn into_pair(self) -> (AccessRelatedness, AccessRelatedness) {
-        match self {
-            AccessRel2XChildY::ChildBoth =>
-                (AccessRelatedness::StrictChildAccess, AccessRelatedness::StrictChildAccess),
-            AccessRel2XChildY::ChildY =>
-                (AccessRelatedness::AncestorAccess, AccessRelatedness::StrictChildAccess),
-            AccessRel2XChildY::ChildNeither =>
-                (AccessRelatedness::DistantAccess, AccessRelatedness::DistantAccess),
+impl PtrSelector {
+    fn rel_pair(self, xy_rel: RelPosXY) -> (AccessRelatedness, AccessRelatedness) {
+        match xy_rel {
+            RelPosXY::MutuallyForeign =>
+                match self {
+                    PtrSelector::X =>
+                        (AccessRelatedness::StrictChildAccess, AccessRelatedness::DistantAccess),
```

Comment:
Why not `This`? Seems more logical -- you are asking how x is related to `self` which is `X`.